File: mcp-server/providers/bohrium.py
"""Bohrium LKM provider backed by the official ``bohr`` CLI."""
from __future__ import annotations
import json
import os
import shutil
import subprocess
from pathlib import Path
from typing import Any

BOHR_TIMEOUT = 30

def _call(args: list[str], keep_error_data: bool = False) -> dict[str, Any]:
    billable = any(part in {"search", "reasoning", "graph", "result"} for part in args)
    if billable and os.environ.get("OSP_ALLOW_LKM_SPEND") != "1":
        return {"error": "LKM spending is not authorized for this run; pass --allow-lkm-spend to osp review"}
    if shutil.which("bohr") is None:
        return {"error": "bohr CLI not found; install @dptech-corp/bohr-cli and run bohr auth login"}
    try:
        result = subprocess.run(["bohr", *args, "--yes", "-o", "json"], capture_output=True, text=True, timeout=BOHR_TIMEOUT, check=False)
        output = result.stdout.strip()
        try:
            payload = json.loads(output)
        except json.JSONDecodeError:
            try:
                payload = next(json.loads(line) for line in reversed(output.splitlines()) if line.strip().startswith("{"))
            except (StopIteration, json.JSONDecodeError) as error:
                return {"error": f"bohr returned invalid JSON: {error}"}
    except subprocess.TimeoutExpired:
        return {"error": f"bohr command timed out after {BOHR_TIMEOUT}s"}
    except (json.JSONDecodeError, IndexError) as error:
        return {"error": f"bohr returned invalid JSON: {error}"}
    if not payload.get("ok", False):
        if keep_error_data and payload.get("data"):
            return payload["data"]
        error = payload.get("error") or {}
        return {"error": error.get("message") or error.get("code") or str(error)}
    return payload.get("data") or {}

def _paper(value: dict[str, Any]) -> dict[str, Any]:
    return {"id": value.get("id", value.get("paperId", "")), "title": value.get("en_title") or value.get("zh_title") or value.get("enName") or value.get("zhName") or "", "authors": value.get("authors") or "", "abstract": value.get("enAbstract") or "", "doi": value.get("doi") or "", "venue": value.get("publication_name") or value.get("publicationEnName") or "", "date": value.get("cover_date_start") or value.get("coverDateStart") or "", "citations": value.get("citationNums", 0), "jcr_zone": value.get("jcrZone") or "", "url": value.get("paperUrl") or ""}

def search_lkm(query: str, top_k: int = 10, scopes: str = "claim,conclusion,abstract") -> dict[str, Any]:
    data = _call(["lkm", "search", query, "--top-k", str(top_k), "--scopes", scopes])
    if "error" in data: return data
    return {"papers": [_paper(item) for item in (data.get("papers") or {}).values()], "variables": (data.get("variables") or [])[:top_k]}

def search_reasoning(query: str, top_k: int = 10) -> dict[str, Any]:
    data = _call(["lkm", "reasoning", "--query", query, "--top-k", str(top_k)])
    if "error" in data: return data
    return {"papers": [_paper(item) for item in (data.get("papers") or {}).values()], "reasoning_chains": (data.get("reasoning_chains") or [])[:top_k]}

def search_papers(query: str, size: int = 10, year_from: int | None = None, year_to: int | None = None, jcr: str | None = None) -> dict[str, Any]:
    args = ["paper", "search", query, "--size", str(size), "--type", "0"]
    if year_from is not None: args += ["--year-from", str(year_from)]
    if year_to is not None: args += ["--year-to", str(year_to)]
    if jcr: args += ["--jcr", jcr]
    data = _call(args)
    if "error" in data: return data
    return {"papers": [_paper(item) for item in data.get("items", [])], "pagination": data.get("pagination") or {}}

def get_paper_graph(paper_id: str, max_nodes: int = 25, max_edges: int = 40) -> dict[str, Any]:
    data = _call(["lkm", "graph", "--paper-id", paper_id], keep_error_data=True)
    if "error" in data: return data
    item = (data.get("items") or [None])[0]
    if not item: return {"error": f"no graph found for paper {paper_id}"}
    graph = item.get("graph") or {}
    nodes, edges = graph.get("nodes") or [], graph.get("edges") or []
    return {"paper": _paper(item.get("paper") or {}), "addressed_problems": item.get("addressed_problems") or [], "open_questions": item.get("open_questions") or [], "graph": {"graph_empty": not nodes, "node_count": len(nodes), "edge_count": len(edges), "nodes": nodes[:max_nodes], "edges": edges[:max_edges]}}

def parse_submit(pdf_path: str) -> dict[str, Any]:
    workspace = os.environ.get("OSP_WORKSPACE_ROOT")
    if not workspace:
        return {"error": "OSP_WORKSPACE_ROOT is not configured"}
    root = (Path(workspace) / ".brain" / "input").resolve()
    candidate = Path(pdf_path).resolve()
    if candidate.parent != root or candidate.suffix.lower() != ".pdf" or not candidate.is_file():
        return {"error": "pdf_path must be a regular PDF directly under .brain/input"}
    if candidate.stat().st_size > 64 * 1024 * 1024:
        return {"error": "PDF exceeds the 64 MiB extraction limit"}
    return _call(["lkm", "parse", "submit", str(candidate)])
def parse_status(task_id: str) -> dict[str, Any]: return _call(["lkm", "parse", "status", task_id])
def parse_wait(task_id: str, interval_s: int = 5, timeout_s: int = 60) -> dict[str, Any]:
    result = _call(["lkm", "parse", "wait", task_id, "--interval", f"{interval_s}s", "--timeout", f"{timeout_s}s"])
    if "error" in result and ("TIMEOUT" in result["error"] or "stopped waiting" in result["error"]):
        return {"task_id": task_id, "status": "running", "warning": result["error"]}
    return result
def parse_result(task_id: str) -> dict[str, Any]: return _call(["lkm", "parse", "result", task_id, "--format", "graph"])

File: src/mcp/providers/bohrium.ts
// Bohrium LKM provider backed by the official `bohr` CLI.
//
// Two safety properties are load-bearing and must not be relaxed:
//   1. Billable subcommands refuse to run unless OSP_ALLOW_LKM_SPEND is exactly
//      "1", so a run without --allow-lkm-spend can never incur a charge.
//   2. parse_submit only accepts a PDF sitting directly in the run's
//      .brain/input, so a prompt cannot walk the agent out of the workspace.
import { execa } from "execa";
import { realpath, stat } from "node:fs/promises";
import { dirname, extname, resolve } from "node:path";

const BOHR_TIMEOUT_MS = 30_000;
const PDFINFO_TIMEOUT_MS = 10_000;
const MAX_PDF_BYTES = 64 * 1024 * 1024;
const MAX_PDF_PAGES = 50;
const BILLABLE = new Set(["search", "reasoning", "graph", "result"]);

export type Json = Record<string, unknown>;
export type Failure = { error: string };

function isFailure(value: Json | Failure): value is Failure {
  return typeof (value as Failure).error === "string";
}

async function pdfPageCount(pdfPath: string): Promise<number | string> {
  try {
    const result = await execa("pdfinfo", [pdfPath], {
      timeout: PDFINFO_TIMEOUT_MS,
      reject: false,
    });
    if (result.failed && result.timedOut) {
      return "pdfinfo timed out while validating the LKM PDF page limit";
    }
    if (result.exitCode !== 0) {
      return "could not determine the PDF page count for LKM extraction";
    }
    const match = /^Pages:\s*(\d+)\s*$/m.exec(result.stdout);
    return match ? Number.parseInt(match[1], 10) : "could not determine the PDF page count for LKM extraction";
  } catch (error) {
    if ((error as { code?: string }).code === "ENOENT") {
      return "pdfinfo is required to validate the LKM PDF page limit";
    }
    return "could not determine the PDF page count for LKM extraction";
  }
}

async function call(args: string[], keepErrorData = false): Promise<Json | Failure> {
  if (args.some((part) => BILLABLE.has(part)) && process.env.OSP_ALLOW_LKM_SPEND !== "1") {
    return { error: "LKM spending is not authorized for this run; pass --allow-lkm-spend to osp review" };
  }

  let stdout: string;
  try {
    const result = await execa("bohr", [...args, "--yes", "-o", "json"], {
      timeout: BOHR_TIMEOUT_MS,
      reject: false,
    });
    if (result.timedOut) return { error: `bohr command timed out after ${BOHR_TIMEOUT_MS / 1000}s` };
    stdout = (result.stdout ?? "").trim();
  } catch (error) {
    if ((error as { code?: string }).code === "ENOENT") {
      return { error: "bohr CLI not found; install @dptech-corp/bohr-cli and run bohr auth login" };
    }
    return { error: `bohr invocation failed: ${(error as Error).message}` };
  }

  let payload: Json;
  try {
    payload = JSON.parse(stdout) as Json;
  } catch {
    // The CLI sometimes prefixes progress lines; the last JSON object wins.
    const line = stdout.split("\n").reverse().find((candidate) => candidate.trim().startsWith("{"));
    if (!line) return { error: "bohr returned invalid JSON: no JSON object in output" };
    try {
      payload = JSON.parse(line) as Json;
    } catch (error) {
      return { error: `bohr returned invalid JSON: ${(error as Error).message}` };
    }
  }

  if (payload.ok !== true) {
    if (keepErrorData && payload.data) return payload.data as Json;
    const error = (payload.error ?? {}) as Json;
    return { error: String(error.message ?? error.code ?? JSON.stringify(error)) };
  }
  return (payload.data as Json) ?? {};
}

function paper(value: Json): Json {
  return {
    id: value.id ?? value.paperId ?? "",
    title: value.en_title || value.zh_title || value.enName || value.zhName || "",
    authors: value.authors ?? "",
    abstract: value.enAbstract ?? "",
    doi: value.doi ?? "",
    venue: value.publication_name || value.publicationEnName || "",
    date: value.cover_date_start || value.coverDateStart || "",
    citations: value.citationNums ?? 0,
    jcr_zone: value.jcrZone ?? "",
    url: value.paperUrl ?? "",
  };
}

function papersFromMap(value: unknown): Json[] {
  return Object.values((value as Record<string, Json>) ?? {}).map(paper);
}

export async function searchLkm(query: string, topK = 10, scopes = "claim,conclusion,abstract"): Promise<Json | Failure> {
  const data = await call(["lkm", "search", query, "--top-k", String(topK), "--scopes", scopes]);
  if (isFailure(data)) return data;
  return { papers: papersFromMap(data.papers), variables: ((data.variables as unknown[]) ?? []).slice(0, topK) };
}

export async function searchReasoning(query: string, topK = 10): Promise<Json | Failure> {
  const data = await call(["lkm", "reasoning", "--query", query, "--top-k", String(topK)]);
  if (isFailure(data)) return data;
  return {
    papers: papersFromMap(data.papers),
    reasoning_chains: ((data.reasoning_chains as unknown[]) ?? []).slice(0, topK),
  };
}

export async function searchPapers(
  query: string,
  size = 10,
  yearFrom?: number,
  yearTo?: number,
  jcr?: string,
): Promise<Json | Failure> {
  const args = ["paper", "search", query, "--size", String(size), "--type", "0"];
  if (yearFrom !== undefined) args.push("--year-from", String(yearFrom));
  if (yearTo !== undefined) args.push("--year-to", String(yearTo));
  if (jcr) args.push("--jcr", jcr);
  const data = await call(args);
  if (isFailure(data)) return data;
  return {
    papers: ((data.items as Json[]) ?? []).map(paper),
    pagination: (data.pagination as Json) ?? {},
  };
}

export async function getPaperGraph(paperId: string, maxNodes = 25, maxEdges = 40): Promise<Json | Failure> {
  const data = await call(["lkm", "graph", "--paper-id", paperId], true);
  if (isFailure(data)) return data;
  const item = ((data.items as Json[]) ?? [])[0];
  if (!item) return { error: `no graph found for paper ${paperId}` };
  const graph = (item.graph as Json) ?? {};
  const nodes = (graph.nodes as unknown[]) ?? [];
  const edges = (graph.edges as unknown[]) ?? [];
  return {
    paper: paper((item.paper as Json) ?? {}),
    addressed_problems: item.addressed_problems ?? [],
    open_questions: item.open_questions ?? [],
    graph: {
      graph_empty: nodes.length === 0,
      node_count: nodes.length,
      edge_count: edges.length,
      nodes: nodes.slice(0, maxNodes),
      edges: edges.slice(0, maxEdges),
    },
  };
}

export async function parseSubmit(pdfPath: string): Promise<Json | Failure> {
  const workspace = process.env.OSP_WORKSPACE_ROOT;
  if (!workspace) return { error: "OSP_WORKSPACE_ROOT is not configured" };

  // Resolve symlinks on both sides before comparing, so a link planted inside
  // .brain/input cannot point at a file outside the workspace.
  const root = await realpath(resolve(workspace, ".brain", "input")).catch(() => null);
  const candidate = await realpath(resolve(pdfPath)).catch(() => null);
  if (!root || !candidate || dirname(candidate) !== root || extname(candidate).toLowerCase() !== ".pdf") {
    return { error: "pdf_path must be a regular PDF directly under .brain/input" };
  }
  const info = await stat(candidate).catch(() => null);
  if (!info?.isFile()) return { error: "pdf_path must be a regular PDF directly under .brain/input" };
  if (info.size > MAX_PDF_BYTES) return { error: "PDF exceeds the 64 MiB extraction limit" };

  const pages = await pdfPageCount(candidate);
  if (typeof pages === "string") return { error: pages };
  if (pages > MAX_PDF_PAGES) return { error: `PDF exceeds the ${MAX_PDF_PAGES}-page LKM extraction limit` };

  return call(["lkm", "parse", "submit", candidate]);
}

export async function parseStatus(taskId: string): Promise<Json | Failure> {
  return call(["lkm", "parse", "status", taskId]);
}

export async function parseWait(taskId: string, intervalS = 5, timeoutS = 60): Promise<Json | Failure> {
  const result = await call(["lkm", "parse", "wait", taskId, "--interval", `${intervalS}s`, "--timeout", `${timeoutS}s`]);
  if (isFailure(result) && (result.error.includes("TIMEOUT") || result.error.includes("stopped waiting"))) {
    // A wait timeout is not a failure: the extraction is still running.
    return { task_id: taskId, status: "running", warning: result.error };
  }
  return result;
}

export async function parseResult(taskId: string): Promise<Json | Failure> {
  return call(["lkm", "parse", "result", taskId, "--format", "graph"]);
}
